authorizer/app.py

```diff
@@ -0,0 +1,364 @@
+
+    if request.method == "POST" and form.validate():
+        new_capabilities = []
+        for capability in capabilities:
+            if form[capability].data:
+                new_capabilities.append(capability)
+        scope = " ".join(new_capabilities)
+        aud = current_app.config.get("OAUTH2_JWT.AUD.DEFAULT", decoded_token["aud"])
+        new_token: Dict[str, Any] = {"scope": scope, "aud": aud}
+        email = decoded_token.get("email")
+        user = decoded_token.get(current_app.config["JWT_USERNAME_KEY"])
+        uid = decoded_token.get(current_app.config["JWT_UID_KEY"])
+        if email:
+            new_token["email"] = email
+        if user:
+            new_token[current_app.config["JWT_USERNAME_KEY"]] = user
+        if uid:
+            new_token[current_app.config["JWT_UID_KEY"]] = uid
+
+        # FIXME: Copies groups. Useful for WebDAV, maybe not necessary
+        #
+        # new_token['isMemberOf'] = decoded_token['isMemberOf']
+        oauth2_proxy_ticket = Ticket()
+        ticket_prefix = current_app.config["OAUTH2_STORE_SESSION"]["TICKET_PREFIX"]
+        ticket_handle = oauth2_proxy_ticket.as_handle(ticket_prefix)
+        new_token["jti"] = ticket_handle
+
+        exp = datetime.utcnow() + timedelta(seconds=current_app.config["OAUTH2_JWT_EXP"])
+        _ = issue_token(
+            new_token, exp=exp, store_user_info=True, oauth2_proxy_ticket=oauth2_proxy_ticket
+        )
+
+        flash(
+            f"Your Newly Created Token. Keep these Secret!<br>\n"
+            f"Token: {oauth2_proxy_ticket} <br>"
+        )
+        return redirect(url_for("tokens"))
+
+    return render_template(
+        "new_token.html", title="New Token", form=form, capabilities=capabilities
+    )
+
+
+def _make_capability_headers(response: Response, encoded_token: str) -> None:
+    """Set Headers scope headers that can be returned in the case of
+    API authorization failure due to required capabiliites.
+    :return: The mutated response object.
+    """
+    decoded_token = jwt.decode(encoded_token, verify=False)
+    capabilities_required, satisfy = verify_authorization_strategy()
+    group_capabilities_set = capabilities_from_groups(decoded_token)
+    scope_capabilities_set = set(decoded_token.get("scope", "").split(" "))
+    user_capabilities_set = group_capabilities_set.union(scope_capabilities_set)
+    response.headers["X-Auth-Request-Token-Capabilities"] = " ".join(user_capabilities_set)
+    response.headers["X-Auth-Request-Capabilities-Accepted"] = " ".join(capabilities_required)
+    response.headers["X-Auth-Request-Capabilities-Satisfy"] = satisfy
+
+
+def _make_success_headers(response: Response, encoded_token: str) -> None:
+    """Set Headers that will be returned in a successful response.
+    :return: The mutated response object.
+    """
+    _make_capability_headers(response, encoded_token)
+
+    decoded_token = jwt.decode(encoded_token, verify=False)
```

Comment:
Why are we not verifying here?  Is it that we assume it's already been done?

Reply:
yeah, `_make_success_headers` should only be called if we are successful with authentication. 

I can add an extra arg to the function which accepts the verified token _also_ instead of decoding it again (same with `_make_capability_headers`). I think the lack of that argument is mostly a relic from before we were reissuing (and just passing the encoded token along)